File: algorithms/insertion_sort.py
""" Insertion Sort """


def insertion_sort(array: list) -> list:
    """
    Implements the insertion sort algorithm
    """
    # [2,5,4,7,1,2]
    for i in range(1, len(array)):
        while i > 0 and array[i - 1] > array[i]:
            temp = array[i]
            array[i] = array[i - 1]
            array[i - 1] = temp
            i -= 1


def main():
    """
    Entrypoint to the program
    """
    array = [5, 1, 3, 2, 3, 6, 8, 2, 4, 8, 90, 2, 3]
    print(array)
    insertion_sort(array)
    print(array)


if __name__ == "__main__":
    main()

File: algorithms/golang/insertion_sort.go
package main

import "fmt"

// insertionSort implements the insertion sort algorithm on an array of integers
func insertionSort(array []int) {
	for i := 1; i < len(array); i++ {
		for i > 0 && array[i] < array[i-1] {
			temp := array[i]
			array[i] = array[i-1]
			array[i-1] = temp
			i--
		}
	}
}

func main() {
	array := []int{4, 5, 2, 1, 3, 6, 7, 8}
	println(fmt.Sprintf("%v", array))
	insertionSort(array)
	println(fmt.Sprintf("%v", array))
}
